### PR TITLE
Closes #2708 derive_vars_dtm function date issue

### LINKED ISSUE #2708: Bug: derive_vars_dtm() always returns max_date() as imputed date when max_date()<min_date()
### What happened?

When working on date imputations we have two variables for max_date and min_date. If max_date is ever less than min_date it overrides the min_date and the imputation will always be set to max_date no matter what. For example if max_date<[date for imputation]<min_date or [date for imputation]<max_date<min_date, it will always impute for max_date, disregarding other rules. Instead what was wanted was to impute to the "first" of the month or year as that is what was specified. 

Looking through the admiral source code, if max_date<min_date<IMPDATE, or max_date<IMPDATE<min_date, first it would do the check for min_date, taking the max of min_date and IMPDATE (call this NEWDATE), next it would do the check for max_date, taking the min of NEWDATE and max_date, which would always equal max_date, overriding the imputation. 

A solution I can think of is setting some sort of priority date, like give either max_date or min_date higher priority if max_date<min_date. Another solution may be comparing max_date and min_date before comparing them to imp_date, instead of comparing imp_date to both sequentially and causing it to get overriden. 

### Session Information

```{r}
R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] logrx_0.3.0     xportr_0.3.2    lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1  
 [6] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1    ggplot2_3.5.0  
[11] tidyverse_2.0.0 metacore_0.1.2  metatools_0.1.5 dplyr_1.1.4     admiral_1.0.2  
[16] haven_2.5.4    

loaded via a namespace (and not attached):
 [1] utf8_1.2.4        generics_0.1.3    xml2_1.3.6        stringi_1.8.3     digest_0.6.34    
 [6] hms_1.1.3         magrittr_2.0.3    grid_4.3.2        timechange_0.3.0  fastmap_1.1.1    
[11] cellranger_1.1.0  sessioninfo_1.2.2 promises_1.2.1    admiraldev_1.0.0  fansi_1.0.6      
[16] scales_1.3.0      cli_3.6.2         shiny_1.8.0       crayon_1.5.2      rlang_1.1.3      
[21] ellipsis_0.3.2    munsell_0.5.0     withr_3.0.0       tools_4.3.2       tzdb_0.4.0       
[26] colorspace_2.1-0  httpuv_1.6.14     waiter_0.2.5      mime_0.12         vctrs_0.6.5      
[31] R6_2.5.1          lifecycle_1.0.4   miniUI_0.1.1.1    pkgconfig_2.0.3   later_1.3.2      
[36] pillar_1.9.0      gtable_0.3.4      Rcpp_1.0.12       glue_1.7.0        tidyselect_1.2.0 
[41] rstudioapi_0.15.0 xtable_1.8-4      htmltools_0.5.7   compiler_4.3.2    readxl_1.4.3
```

### Reproducible Example

```{r} 

AESTDTC <- c("2022-01", "2022-01", "2022-01", "2022-01", "2022-01", "2022-01", "2022-11") 

AEENDTC <- c("2021-12-31", "2022-01-30", "2022-02-28", "2022-11-05", "2022-11-30", "2022-12-31", "2022-11-02") 

ae <- data.frame(AESTDTC, AEENDTC) %>% 
 mutate(TRTSDT = as_date("2022-11-08"), 
  AEENDTCdate = as_date(AEENDTC)) %>% 
derive_vars_dtm(dtc = AESTDTC, 
 new_vars_prefix = "AST", 
 highest_imputation="M", 
 date_imputation = "first", 
 min_dates = exprs(TRTSDT), 
 max_dates = exprs(AEENDTCdate), 
 flag_imputation="date" ) 

``` 

 

In the above code we are using AESTDTC as the original dtc var being converted to dt format, max_date is the date portion of AEENDTC if it exists, and min_date is TRTSDT. We want to impute to the first value (ie first of the month or TRTSDT, whichever is later) 

On line 7 of the output dataset we can see that ASTDTM, where it should have been done to the first of the month, instead be set to max_date. 

Image

### PR DESCRIPTION
**Pull request by sadchla-codes**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/cran-release/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF and a PNG version of it to the same folder. (The PNG version can be created by taking a screenshot of the PDF version.)
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by sadchla-codes** [comment]
@bms63 @manciniedoardo which is the correct branch to use as base
image

**Comment by bms63** [comment]
main is what we do PRs into

**Comment by bms63** [comment]
Hey @sadchla-codes thanks for jumping on this.  I didn't look at code just yet and it looked invovled.

Can you write some tests for this scenario being addressed with your update- I think the #2708  has an example that could be incorporated so we can make sure the update to the function is doing what is intended.  TYTY

**Comment by sadchla-codes** [comment]
@bms63 test scenarios have been added

File: tests/testthat/test-derive_vars_dtm.R [hunk 1]
```diff
@@ -891,3 +891,152 @@ test_that("derive_vars_dt Test 30: impute_dtc_dt where dtc is empty", {
...
+  TRTSDT <- c( # nolint
+    as.Date("2021-12-15"),
+    as.Date("2022-01-10")
+  )
+
+  ae <- data.frame(AESTDTC, AEENDTC, stringsAsFactors = FALSE) %>%
+    mutate(
+      TRTSDT      = TRTSDT,
+      AEENDTCdate = as.Date(AEENDTC)
+    ) %>%
```
Comment (bms63) [inline tests/testthat/test-derive_vars_dtm.R:965]: can we follow admiral conventions for making dummy test data for a test and have it all combined in one dataframe/tibble.

Also what is AEENDTCdate doing here?

Comment (bms63) [reply tests/testthat/test-derive_vars_dtm.R:965]: i believe if you use tibble you won't need stringAsFactors either

**Comment by bms63** [comment]
> @bms63 test scenarios have been added

awesome thanks!  can we **admiralize** the input data for the tests please?

**Comment by Kcdahn** [comment]
Add some user  parameters?
(1) Planned Treatment Duration:  Adjustment (add) For when a xxTPT is  END / AFTER Treatment etc. as a tablet or injection study might 0 hours, infusion study could be 30 Mins, 1 Hour, etc.
(2) Added parameter for  Planned Screen Run-in hours:  For when the xxTPT is screening. Subtract from Zero adjustment. 
(3) Add a range handling parameter (BEG, MID, END):   If a xxTPT is a range (ex  4-8 Hours Post Dose).  BEG would return 4 Hours, END would return the 8.  MID would return 6 (computes the mid point).

**Comment by bms63** [comment]
> Add some user parameters? (1) Planned Treatment Duration: Adjustment (add) For when a xxTPT is END / AFTER Treatment etc. as a tablet or injection study might 0 hours, infusion study could be 30 Mins, 1 Hour, etc. (2) Added parameter for Planned Screen Run-in hours: For when the xxTPT is screening. Subtract from Zero adjustment. (3) Add a range handling parameter (BEG, MID, END): If a xxTPT is a range (ex 4-8 Hours Post Dose). BEG would return 4 Hours, END would return the 8. MID would return 6 (computes the mid point).

I'm happy for this to be made into an issue and explore - but for this Pull Request/Issue we should stick within the scope or else we can run the risk of never finishing this update.

**Comment by manciniedoardo** [comment]
> Add some user parameters? (1) Planned Treatment Duration: Adjustment (add) For when a xxTPT is END / AFTER Treatment etc. as a tablet or injection study might 0 hours, infusion study could be 30 Mins, 1 Hour, etc. (2) Added parameter for Planned Screen Run-in hours: For when the xxTPT is screening. Subtract from Zero adjustment. (3) Add a range handling parameter (BEG, MID, END): If a xxTPT is a range (ex 4-8 Hours Post Dose). BEG would return 4 Hours, END would return the 8. MID would return 6 (computes the mid point).

Is this meant for #2917 ?

**Comment by bms63** [comment]
Hey @sadchla-codes I was wondering why you didn't use the suggest arguments in the issue.  @bundfussr and I had a conversation about them later in the issue.  Could you look into implementing those please...that might give a little more freedom to how the function works.

Stefan's is also going to provide some additional feedback on some of the code soon.

Thanks again for doing this - but this is a stressful change to this function and we are going to implement in 1.5 as this one will be in review for a while.

File: R/derive_vars_dtm.R [hunk 2]
```diff
@@ -655,58 +655,108 @@ restrict_imputed_dtc_dtm <- function(dtc,
...
+
+  ## ---- Handle inconsistent bounds: max < min ---------------------------
+  # If both bounds are present but max < min, treat them as inconsistent
+  # and fall back to the original imputed value (ignore both bounds).
+  inconsistent <- !is.na(min_bound) & !is.na(max_bound) & max_bound < min_bound
+  if (any(inconsistent)) {
+    # Optional: could cli_inform() here if you want a warning
+    min_bound[inconsistent] <- NA_character_
+    max_bound[inconsistent] <- NA_character_
+  }
```
Comment (bundfussr) [inline R/derive_vars_dtm.R:735]: This is not the expected behavior. Consider for example imputation of `ASTDTM` where `min_dates = TRTSDTM` and `max_dates = AENDTM`. Both dates should be considered. To handle it correctly two new arguments `min_dates_strict` and `max_dates_strict` should be implemented (see https://github.com/pharmaverse/admiral/issues/2708#issuecomment-2825027344 for details). In the example above `min_dates = TRTSDTM` and `max_dates_strict = AENDTM` would be used.

**Comment by manciniedoardo** [comment]
Thanks  @sadchla-codes for working on this. We discussed this issue in the deep dive meeting today (9th feb) and feel the problem is quite subtle and complex and would benefit from a more well-defined issue. So, we will close this PR, and in the next deep dive meet (23rd Feb) we will review issue #2708 to summarise what we need to add/change, perhaps making a new issue too.
